lib/guard.js

```diff
@@ -0,0 +1,140 @@
+
+        Object.defineProperty(this, 'id', {
+            value: id || crypto.randomBytes(3 * 4).toString('base64'),
+            enumerable: true,
+        });
+
+        Object.defineProperty(this, 'metricsThreshold', {
+            value: metricsThreshold,
+        });
+
+        Object.defineProperty(this, 'permutationWarnThreshold', {
+            value: Math.floor((permutationThreshold / 100) * 80),
+        });
+
+        Object.defineProperty(this, 'permutationThreshold', {
+            value: permutationThreshold,
+        });
+
+        Object.defineProperty(this, 'enabled', {
+            value: enabled,
+        });
+
+        Object.defineProperty(this, 'registry', {
+            value: new Map(),
+        });
+    }
+
+    get [Symbol.toStringTag]() {
+        return 'MetricsGuard';
+    }
+
+    [push](metric) {
+        const met = metric;
+        met.source = this.id;
+        // eslint-disable-next-line no-underscore-dangle
+        if (this._readableState.flowing) {
+            this.push(met);
+            return;
+        }
+        this.emit('drop', met);
+    }
+
+    getMetrics() {
+        return Array.from(this.registry.keys());
+    }
+
+    getLabels(metric) {
+        const labels = this.registry.get(metric);
+        if (labels) {
+            return Array.from(labels.keys());
+        }
+        return [];
+    }
+
+    reset() {
+        this.registry.clear();
+    }
+
+    _transform(metric, enc, next) {
+        // guarding is disabeled, pass metrich through
+        if (!this.enabled) {
+            next(null, metric);
+            return;
+        }
+
+        // number of metrics registered exceeds the threshold
+        // emit warning
+        if (this.registry.size >= this.metricsThreshold) {
+            this.emit('warn', 'metrics', this.registry.size);
+        }
+
+        let entry = this.registry.get(metric.name);
+
+        if (entry) {
+            // too many label values is registered on the metric
+            // drop the metric
+            if (entry.size >= this.permutationThreshold) {
+                this.emit('drop', metric);
+                next(null);
+                return;
+            }
+
+            // getting close to the threshold of too many label
+            // values registered on the metric so warn
+            if (entry.size >= this.permutationWarnThreshold) {
+                this.emit('warn', 'permutation', metric.name);
+            }
+        } else {
+            // metric was not in registry, greate a new entry in
+            // the registry
+            entry = new Set();
+            this.registry.set(metric.name, entry);
+        }
+
+        // push all label values on the metric into the Set
+        // for the metric in the registry. the size of this
+        // Set is compared with the threshold
+        metric.labels.forEach((label) => {
+            entry.add(label.value);
+        });
```

Comment:
The size of the `entry` set could continue growing even if metrics are being dropped?

Reply:
It will grow until one start droping metrics. When reaching the threshold for dropping here, new values will not be set so this will not grow.